fake.go

```diff
@@ -0,0 +1,161 @@
+//        fc.AddCall("Send", FakeCallArgs{
+//            "request": request,
+//        })
+//        return fc.Response, fc.Err()
+//    }
+//
+// As demonstrated in the example, embed a pointer to testing.Fake. This
+// allows a single testing.Fake to be shared between multiple fakes.
+//
+// Error return values are set through Fake.Errors. Set it to the errors
+// you want returned. The errors will be matched up to the calls made on
+// Fake methods, in order. This is facilitated by the Err method, as
+// seen in the above example. In some cases the first method call is not
+// the one you want to fail. If not then put a nil before the error in
+// Fake.Errors. Fake.SetErrors is a helper for setting up failure cases.
+//
+// To validate calls made to the fake in a test, call the CheckCalls
+// method:
+//
+//    s.fake.CheckCalls(c, []FakeCall{{
+//        FuncName: "Send",
+//        Args: FakeCallArgs{
+//            "request": expected,
+//        },
+//    }
+//
+// Not only is Fake useful for building a interface implementation to
+// use in testing (e.g. a network API client), it is also useful in
+// patching situations:
+//
+//    s.PatchValue(&somefunc, s.fake.Send)
+//
+// This allows for easily monitoring the args passed to the patched
+// func, as well as controlling the return value from the func in a
+// clean manner (by simply setting the correct fake field).
+type Fake struct {
+	// Calls is the list of calls that have been registered on the fake
+	// (i.e. made on the fake's methods), in the order that they were
+	// made.
+	Calls []FakeCall
+
+	// Errors holds the list of error return values to use for
+	// successive calls to methods that return an error. Each call
+	// pops the next error off the list. An empty list (the default)
+	// implies a nil error. nil may be precede actual errors in the
+	// list, which means that the first calls will succeed, followed
+	// by the failure. All this is facilitated through the Err method.
+	Errors []error
+
+	// Error is the default error (when Errors is empty). The typical
+	// Fake usage will leave this nil (i.e. no error).
+	Error error
+}
+
+// Error returns the error that should be returned on the nth call to
+// any method on the fake. It should be called for the error return in
+// all faked methods.
+func (f *Fake) Err() error {
+	if len(f.Errors) == 0 {
+		return f.Error
+	}
+	err := f.Errors[0]
+	f.Errors = f.Errors[1:]
+	return err
+}
+
+// AddCall records a faked method call for later inspection using the
+// CheckCalls method. All faked methods should call AddCall.
+func (f *Fake) AddCall(funcName string, args FakeCallArgs) {
+	f.Calls = append(f.Calls, FakeCall{
+		FuncName: funcName,
+		Args:     args,
+	})
+}
+
+// AddRcvrCall records a faked method call for later inspection using
+// the CheckCalls method. All faked methods should call AddCall.
+func (f *Fake) AddRcvrCall(receiver interface{}, funcName string, args FakeCallArgs) {
```

Comment:
This doesn't really seem useful....  if the receiver is a pointer (which they often are), then it could change between when you call this and when you check it later.  I guess if you're careful to never pass a pointer to this method (even if the receiver really should be a pointer), then it's ok.  But it seems like an annoying gotcha (not that I can really think of a better way to do it).